### ISSUE TITLE
agda --interaction-json doesn't give ranges for invisible goals

### ISSUE THREAD

**Issue by msuperdock** [agda/agda#4823]
## Issue

In emacs, the `C-c C-l` command gives ranges for invisible goals, which is nice because these invisible goals can otherwise be hard to find. However, when using `--interaction-json`, we get no such ranges.

## Example

For example, consider the following file:

```
/data/code/agda-test/Test.agda
--
module Test where

test = ?
```

In emacs, the `C-c C-l` command gives:

```
?0 : _1
Sort _0  [ at /data/code/agda-test/Test.agda:3,1-5 ]
_1 : _0  [ at /data/code/agda-test/Test.agda:3,1-5 ]
```

But the data provided under `agda --interaction-json` is only as follows:

```
{
  "status" : {
    "checked" : false,
    "showImplicitArguments" : false
  },
  "kind" : "Status"
}
{
  "kind" : "DisplayInfo",
  "info" : {
    "visibleGoals" : [
      {
        "constraintObj" : 0,
        "kind" : "OfType",
        "type" : "_1"
      }
    ],
    "kind" : "AllGoalsWarnings",
    "warnings" : "",
    "invisibleGoals" : [
      {
        "constraintObj" : "_0",
        "kind" : "JustSort"
      },
      {
        "constraintObj" : "_1",
        "kind" : "OfType",
        "type" : "_0"
      }
    ],
    "errors" : ""
  }
}
{
  "kind" : "InteractionPoints",
  "interactionPoints" : [
    {
      "range" : [
        {
          "start" : {
            "line" : 3,
            "pos" : 27,
            "col" : 8
          },
          "end" : {
            "line" : 3,
            "pos" : 28,
            "col" : 9
          }
        }
      ],
      "id" : 0
    }
  ]
}
```

## Thoughts

We get position information for the visible goals, but not the invisible goals. This makes it difficult for a non-emacs agda plugin based on `--interaction-json` to replicate the information provided by the emacs agda plugin.

Relevant code points are:
- `Agda.Interaction.BasicOps`, lines 697-700; here the emacs mode uses the TCM monad to get ranges for invisible goals.
- `Agda.Interaction.JSONTop`, lines 143-212; here the JSON value for a goal is computed. (This computation has access to the TCM monad, so it could get the range just as is done for the emacs mode by imitating the code in `BasicOps`.)

It would be nice to modify the JSON values, at least for invisible goals, to include range information. As a side point, I think it would also be nice if the JSON values included the full string output provided by the emacs mode; that way, a non-emacs plugin could easily replicate the emacs output, without redoing the work of converting the JSON values into strings.

**Comment by rwe** [comment]
Maybe relevant to @ice1000 or @banacorn who've contributed in that area before. Should this be tagged as a bug? Or missing feature?

**Comment by ice1000** [comment]
Are you doing some Agda tooling? Can you talk about that?

Speaking of the issue, if you wanna get credit for it, you can try to fix it. If you wish me or @banacorn to fix it, you're gonna wait for some time.

**Comment by msuperdock** [comment]
@rwe: I think the "type: enhancement" label is fair.

@ice1000: I've made an Agda plugin for neovim using `--interaction-json`, and just today updated it to handle the new JSON format used by Agda 2.6.1. I just uploaded it to github; [here](https://github.com/msuperdock/vim-agda) is a link.

I can try adding a field for the range; I just wanted to check first to see if there was some agreement that it would be a good thing to add. It sounds like you agree?

**Comment by rwe** [comment]
Cool! I contributed a bit to the agda-vim repo last year (https://github.com/derekelkins/agda-vim/commits/f928f453f1f48950b3be787419e74b1b90794e43), which you listed as one of the references, though I didn't do much; async interaction would help a lot.

As far as interaction-json issues like this, it might also be interesting to keep in mind ways of anticipating or contributing toward a design Language Server Protocol support. (@L-TChen was working on it in #4359 and I wished for it in #4820; not sure how far that got).

**Comment by msuperdock** [comment]
Closed by #4829.
